README.md
# pyhsslms

HSS/LMS Digital Signature library for Python
--------------------------------------------
[PyPI](https://pypi.org/project/pyhsslms)

[Python Versions](https://pypi.org/project/pyhsslms/)

[GitHub license](https://raw.githubusercontent.com/russhousley/pyhsslms/master/LICENSE.txt)

This Python package contains a free and open source implementation of
HSS/LMS Hash-based Digital Signatures as defined in [RFC 8554](https://www.rfc-editor.org/rfc/rfc8554.txt).

Features
--------

* Generate HSS/LMS private keys and then sign with them
* Validate signatures with HSS/LMS public keys
* 100% Python, works with Python 2.7 and 3.5+

How to use pyhsslms
-------------------

Generate a HSS/LMS private key:

```python
priv_key = pyhsslms.HssLmsPrivateKey.genkey('mykey', levels=2)
```

The private key is stored in mykey.prv, and the public key is
stored in mykey.pub.  Of course, the mykey.prv must be protected
from disclosure, and it gets updated every time a signature is
created.  Restoring mykey.prv from backup can cause a node in the
tree to be used more that once, forfeiting all security.


Sign a file with a HSS/LMS private key:

```python
priv_key.signFile('myfile.txt')
```
The private key was generated above is used to sign the content of
myfile.txt, and the signature is stored in myfile.txt.sig.


Sign a buffer with a HSS/LMS private key:

```python
sigbuf = prv_key.sign(buffer)
```
The private key was generated above is used to sign the content of
buffer, and the signature is returned in sigbuf.


Verify a signature on a file with a HSS/LMS public key:

```python
pub_key = pyhsslms.HssLmsPublicKey('mykey')
if pub_key.verifyFile('myfile.txt'):
    print('Signature is valid')
else:
    print('Signature is NOT valid!')
```


Verify a signature on a buffer with a HSS/LMS public key:

```python
pub_key = pyhsslms.HssLmsPublicKey('mykey')
validity = pub_key.verify(buffer, sigbuf)
if validity:
    print('Signature is valid')
else:
    print('Signature is NOT valid!')
```


How to get pyhsslms
-------------------

The pyhsslms package is distributed under terms and conditions of
[license](https://raw.githubusercontent.com/russhousley/pyhsslms/master/LICENSE.txt).

Source code is freely available as a GitHub [repo](https://github.com/russhousley/pyhsslms).

You could `pip install pyhsslms` or download it from [PyPI](https://pypi.org/project/pyhsslms).

Copyright (c) 2020-2023, Vigil Security, LLC
All rights reserved.


CHANGES.txt

Revision 1.0.0, released 4-May-2020
-----------------------------------

- Initial release.


Revision 1.0.1, released 14-Jul-2020
------------------------------------

- Use os.urandom instead of secrets, which is not available until Python3.6.


Revision 1.0.2, released 8-Apr-2021
------------------------------------

- Prepare for parameter sets that use SHAKE256:
  -- Use pycryptodomex before Python3.6.
  -- Use hashlib for Python3.6 and later.
  -- Add tests for SHA256 and SHAKE256.

- Clean up the random number generation:
  -- Use os.urandom before Python3.6.
  -- Use secrets.token_bytes for Python3.6 and later.


Revision 1.1.0, released 3-Jun-2021
------------------------------------

- Add parameter sets that use SHA-256/192 and SHAKE256 from
  draft-fluhrer-lms-more-parm-sets-05, which includes the
  IANA-assigned identifiers.

- Refactor unit tests that use files to ensure that files are deleted,
  even when the test fail.

- Fix bug in HssLmsPublicKey.verifyFile().


Revision 1.1.1, released 9-Jan-2023
------------------------------------

- Output the maximum number of signatures that can be produced with a
  private key (Contribution by sebastien-riou).
  
- Correct bug related to the --trunc command line argument (Contribution
  by sebastien-riou).

- Make it easier to run test from the command line (Contribution by
  sebastien-riou).


Revision 2.0.0, released 26-Oct-2023
------------------------------------

- HssPrivateKey serialization supports significantly larger tree sizes.
  Previous approach allowed for a total of (2**32)-1 signature.  This
  change does not break compatibility for private keys generated using
  the old serialization scheme (Contribution by joshuayuen99).

- Generate the LMOTS public key from a given LMOTS private key using
  the approach documented in RFC 8554 (Contribution by joshuayuen99).

- Correctly sign messages with LMOTS private keys (Contribution by
  joshuayuen99).

- Correctly calculate the the size of the LMOTS public key during
  deserialization (Contribution by joshuayuen99).

- Refactor "S" argument within LmotsSignature, LmotsPrivateKey, and
  LmotsPublicKey to "I" and "q" arguments to align with RFC 8554
  (Contribution by joshuayuen99).

- Add tests for the coef function.

- Include support for Classifier: Python 3.9, Python 3.10, Python 3.11,
  and Python 3.12.

- Update copyright.


pyhsslms/compat.py
# compat.py
#
# For compatibility between Python2 and Python3 within the routines
# for HSS/LMS Hash-based Signatures as defined in RFC 8554.
#
#
# Copyright (c) 2020-2023, Vigil Security, LLC
# All rights reserved.
#
# Redistribution and use, with or without modification, are permitted
# provided that the following conditions are met:
#
# (1) Redistributions must retain the above copyright notice, this
#     list of conditions, and the following disclaimer.
#
# (2) Redistributions in binary form must reproduce the above
#     copyright notice, this list of conditions and the following
#     disclaimer in the documentation and/or other materials provided
#     with the distribution.
#
# (3) Neither the name of the Vigil Security, LLC nor the names of the
#     contributors to this code may be used to endorse or promote any
#     products derived from this software without specific prior written
#     permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) REGARDLESS OF THE
# CAUSE AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
# WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from sys import version_info

if version_info[0] <= 2 or (version_info[0] == 3 and version_info[1] < 6):
    from Cryptodome.Hash import SHAKE256

    class crypto_hash_shake256():
        """
        Put a wrapper around Crypto.Hash.SHAKE256 with an interface similar
        to hashlib for use with older versions of Python.
        """
        def __init__(self):
            self.name = 'shake_256'
            self.handle = SHAKE256.new()

        def update(self, buf):
            self.handle.update(buf)

        def digest(self, rvlen):
            return self.handle.read(rvlen)


if version_info[0] <= 2:
    import os
    import struct
    import binascii
    
    u32 = lambda i: struct.pack('>I', i)
    u16 = lambda i: struct.pack('>H', i)
    u8 = chr
    fromHex = binascii.unhexlify
    toHex = binascii.hexlify
    toBytes = lambda x: x
    randBytes = os.urandom
    int32 = lambda x: struct.unpack('>L', x)[0]
    charNum = ord
    shake256 = crypto_hash_shake256
    NoFileError = IOError
    FoundFileError = IOError

elif version_info[0] == 3 and version_info[1] < 6:
    import os

    u32 = lambda i: i.to_bytes(4, byteorder='big', signed=False)
    u16 = lambda i: i.to_bytes(2, byteorder='big', signed=False)
    u8 = lambda i: i.to_bytes(1, byteorder='big', signed=False)
    fromHex = bytes.fromhex
    toHex = lambda x: x.hex()
    toBytes = lambda x: x.encode()
    randBytes = os.urandom
    int32 = lambda x: int.from_bytes(x, byteorder='big')
    charNum = lambda x: x
    shake256 = crypto_hash_shake256
    NoFileError = FileNotFoundError
    FoundFileError = FileExistsError

else:
    import hashlib
    from secrets import token_bytes as random_bytes

    u32 = lambda i: i.to_bytes(4, byteorder='big', signed=False)
    u16 = lambda i: i.to_bytes(2, byteorder='big', signed=False)
    u8 = lambda i: i.to_bytes(1, byteorder='big', signed=False)
    fromHex = bytes.fromhex
    toHex = lambda x: x.hex()
    toBytes = lambda x: x.encode()
    randBytes = random_bytes
    int32 = lambda x: int.from_bytes(x, byteorder='big')
    charNum = lambda x: x
    shake256 = hashlib.shake_256
    NoFileError = FileNotFoundError
    FoundFileError = FileExistsError


pyhsslms/__init__.py
# https://www.python.org/dev/peps/pep-0396/
__version__ = '2.0.0'

from .pyhsslms import lmots_sha256_n32_w1
from .pyhsslms import lmots_sha256_n32_w2
from .pyhsslms import lmots_sha256_n32_w4
from .pyhsslms import lmots_sha256_n32_w8
from .pyhsslms import lmots_sha256_n24_w1
from .pyhsslms import lmots_sha256_n24_w2
from .pyhsslms import lmots_sha256_n24_w4
from .pyhsslms import lmots_sha256_n24_w8
from .pyhsslms import lmots_shake_n32_w1
from .pyhsslms import lmots_shake_n32_w2
from .pyhsslms import lmots_shake_n32_w4
from .pyhsslms import lmots_shake_n32_w8
from .pyhsslms import lmots_shake_n24_w1
from .pyhsslms import lmots_shake_n24_w2
from .pyhsslms import lmots_shake_n24_w4
from .pyhsslms import lmots_shake_n24_w8

from .pyhsslms import lms_sha256_m32_h5
from .pyhsslms import lms_sha256_m32_h10
from .pyhsslms import lms_sha256_m32_h15
from .pyhsslms import lms_sha256_m32_h20
from .pyhsslms import lms_sha256_m32_h25
from .pyhsslms import lms_sha256_m24_h5
from .pyhsslms import lms_sha256_m24_h10
from .pyhsslms import lms_sha256_m24_h15
from .pyhsslms import lms_sha256_m24_h20
from .pyhsslms import lms_sha256_m24_h25
from .pyhsslms import lms_shake_m32_h5
from .pyhsslms import lms_shake_m32_h10
from .pyhsslms import lms_shake_m32_h15
from .pyhsslms import lms_shake_m32_h20
from .pyhsslms import lms_shake_m32_h25
from .pyhsslms import lms_shake_m24_h5
from .pyhsslms import lms_shake_m24_h10
from .pyhsslms import lms_shake_m24_h15
from .pyhsslms import lms_shake_m24_h20
from .pyhsslms import lms_shake_m24_h25

from .pyhsslms import coef
from .pyhsslms import checksum
from .pyhsslms import serialize_list

from .pyhsslms import LmotsSignature
from .pyhsslms import LmotsPrivateKey
from .pyhsslms import LmotsPublicKey

from .pyhsslms import LmsSignature
from .pyhsslms import LmsPrivateKey
from .pyhsslms import LmsPublicKey

from .pyhsslms import HssSignature
from .pyhsslms import HssPrivateKey
from .pyhsslms import HssPublicKey

from .pyhsslms import HssLmsSignature
from .pyhsslms import HssLmsPrivateKey
from .pyhsslms import HssLmsPublicKey


pyhsslms/hsslms.py
#!/usr/bin/env python

# hsslms.py
#
# This provides a command line interface for the pyhsslms.py
# implementation of HSS/LMS Hash-based Signatures as defined
# in RFC 8554.
#
#
# Copyright (c) 2020-2023, Vigil Security, LLC
# All rights reserved.
#
# Redistribution and use, with or without modification, are permitted
# provided that the following conditions are met:
#
# (1) Redistributions must retain the above copyright notice, this
#     list of conditions, and the following disclaimer.
#
# (2) Redistributions in binary form must reproduce the above
#     copyright notice, this list of conditions and the following
#     disclaimer in the documentation and/or other materials provided
#     with the distribution.
#
# (3) Neither the name of the Vigil Security, LLC nor the names of the
#     contributors to this code may be used to endorse or promote any
#     products derived from this software without specific prior written
#     permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) REGARDLESS OF THE
# CAUSE AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
# WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import sys
import os.path
import argparse
import pyhsslms
from .__init__ import __version__ as VERSION


def usage(name):
    """
    Display usage information and then exit.
    """
    cmd_name = os.path.basename(name)
    print("commands:")
    print(cmd_name + " genkey <keyname> [<genparms>]")
    print("   creates <keyname>.prv and <keyname>.pub")
    print(" ")
    print(cmd_name + " sign <keyname> <filename>")
    print("   updates <keyname>.prv and makes the signature in <filename>.sig")
    print(" ")
    print(cmd_name + " verify <keyname> <filename>")
    print("   verifies the signature in <filename>.sig with <keyname>.pub")
    print(" ")
    print(cmd_name + " showprv <keyname>")
    print("   display <keyname>.prv")
    print(" ")
    print(cmd_name + " showpub <keyname>")
    print("   display <keyname>.pub")
    print(" ")
    print(cmd_name + " showsig <filename>")
    print("   display <filename>.sig")
    print(" ")
    print("optional <genparms> for the genkey command:")
    print("   -l LEVELS, --levels LEVELS")
    print("                           Number of levels in HSS heirarchy")
    print("   -s LMS_TYPE, --lms LMS_TYPE")
    print("                           Height of the LMS trees")
    print("   -w LMOTS_TYPE, --lmots LMOTS_TYPE")
    print("                           Winternitz number")
    print("   -a HASH_ALG, --alg HASH_ALG")
    print("                           Hash algorithm (sha256 or shake)")
    print("   -t TRUNC, --trunc TRUNC")
    print("                           Hash algorithm truncation size")
    print(" ")
    print("optional command arguments:")
    print("   -h, --help")
    print("                           Provides this information")
    print("   -v, --version")
    print("                           Provids the program version number")
    sys.exit(1)


def main():
    """
    Command line interface for pyhsslms.py.
    """
    cmds = ['genkey', 'keygen', 'sign', 'verify', \
            'showprv', 'showpub', 'showsig', \
            '--version', '-v', 'version', '--help', '-h', 'help']

    if len(sys.argv) < 2 or sys.argv[1] not in cmds:
        print("error: first argument must be a command")
        usage(sys.argv[0])
        sys.exit(1)

    if sys.argv[1] == 'help' or '--help' in sys.argv or '-h' in sys.argv:
        usage(sys.argv[0])
        sys.exit(1)

    if sys.argv[1] == 'version' or '--version' in sys.argv or '-v' in sys.argv:
        print(os.path.basename(sys.argv[0]) + " " + VERSION)
        sys.exit(1)

    if sys.argv[1] in ['genkey', 'keygen']:
        if len(sys.argv) < 3:
            print("error: second argument must be a keyname")
            usage(sys.argv[0])
            sys.exit(1)

        keyname = sys.argv[2]
        levels = 2
        lms_type = pyhsslms.lms_sha256_m32_h5
        lmots_type = pyhsslms.lmots_sha256_n32_w8
        if len(sys.argv) > 3:
            parser = argparse.ArgumentParser()
            parser.add_argument('-l', '--levels', dest='levels', default=2,
                type=int, choices=[1, 2, 3, 4, 5, 6, 7, 8],
                metavar='LEVELS', help='Number of levels in HSS heirarchy')
            parser.add_argument("-s", "--lms", dest='lms', default=5,
                type=int, choices=[5, 10, 15, 20, 25],
                metavar='LMS_TYPE', help='Height of the LMS trees')
            parser.add_argument('-w', '--lmots', dest='lmots', default=8,
                type=int, choices=[1, 2, 4, 8],
                metavar='LMOTS_TYPE', help='Winternitz number')
            parser.add_argument('-a', '--alg', dest='alg', default='sha256',
                type=str, choices=['sha256', 'shake'],
                metavar='HASH_ALG', help='Hash algorithm (sha256 or shake)')
            parser.add_argument('-t', '--trunc', dest='trunc', default='32',
                type=int, choices=[32, 24],
                metavar='TRUNC', help='Hash algorithm truncation size')
            args = parser.parse_args(sys.argv[3:])

            levels = args.levels
            if args.alg == 'sha256':
                if args.trunc == 32:
                     if args.lms ==  5: lms_type = pyhsslms.lms_sha256_m32_h5
                     if args.lms == 10: lms_type = pyhsslms.lms_sha256_m32_h10
                     if args.lms == 15: lms_type = pyhsslms.lms_sha256_m32_h15
                     if args.lms == 20: lms_type = pyhsslms.lms_sha256_m32_h20
                     if args.lms == 25: lms_type = pyhsslms.lms_sha256_m32_h25
                     if args.lmots == 1: lmots_type = pyhsslms.lmots_sha256_n32_w1
                     if args.lmots == 2: lmots_type = pyhsslms.lmots_sha256_n32_w2
                     if args.lmots == 4: lmots_type = pyhsslms.lmots_sha256_n32_w4
                     if args.lmots == 8: lmots_type = pyhsslms.lmots_sha256_n32_w8
                else: # args.trunc == 24
                     if args.lms ==  5: lms_type = pyhsslms.lms_sha256_m24_h5
                     if args.lms == 10: lms_type = pyhsslms.lms_sha256_m24_h10
                     if args.lms == 15: lms_type = pyhsslms.lms_sha256_m24_h15
                     if args.lms == 20: lms_type = pyhsslms.lms_sha256_m24_h20
                     if args.lms == 25: lms_type = pyhsslms.lms_sha256_m24_h25
                     if args.lmots == 1: lmots_type = pyhsslms.lmots_sha256_n24_w1
                     if args.lmots == 2: lmots_type = pyhsslms.lmots_sha256_n24_w2
                     if args.lmots == 4: lmots_type = pyhsslms.lmots_sha256_n24_w4
                     if args.lmots == 8: lmots_type = pyhsslms.lmots_sha256_n24_w8
            else: # args.alg == 'shake'
                if args.trunc == 32:
                     if args.lms ==  5: lms_type = pyhsslms.lms_shake_m32_h5
                     if args.lms == 10: lms_type = pyhsslms.lms_shake_m32_h10
                     if args.lms == 15: lms_type = pyhsslms.lms_shake_m32_h15
                     if args.lms == 20: lms_type = pyhsslms.lms_shake_m32_h20
                     if args.lms == 25: lms_type = pyhsslms.lms_shake_m32_h25
                     if args.lmots == 1: lmots_type = pyhsslms.lmots_shake_n32_w1
                     if args.lmots == 2: lmots_type = pyhsslms.lmots_shake_n32_w2
                     if args.lmots == 4: lmots_type = pyhsslms.lmots_shake_n32_w4
                     if args.lmots == 8: lmots_type = pyhsslms.lmots_shake_n32_w8
                else: # args.trunc == 24
                     if args.lms ==  5: lms_type = pyhsslms.lms_shake_m24_h5
                     if args.lms == 10: lms_type = pyhsslms.lms_shake_m24_h10
                     if args.lms == 15: lms_type = pyhsslms.lms_shake_m24_h15
                     if args.lms == 20: lms_type = pyhsslms.lms_shake_m24_h20
                     if args.lms == 25: lms_type = pyhsslms.lms_shake_m24_h25
                     if args.lmots == 1: lmots_type = pyhsslms.lmots_shake_n24_w1
                     if args.lmots == 2: lmots_type = pyhsslms.lmots_shake_n24_w2
                     if args.lmots == 4: lmots_type = pyhsslms.lmots_shake_n24_w4
                     if args.lmots == 8: lmots_type = pyhsslms.lmots_shake_n24_w8
        
        pyhsslms.HssLmsPrivateKey.genkey(keyname, levels=levels,
            lms_type=lms_type, lmots_type=lmots_type)

    if sys.argv[1] == 'sign':
        if len(sys.argv) < 3:
            print("error: second argument must be a keyname")
            usage(sys.argv[0])
            sys.exit(1)

        if len(sys.argv) < 4:
            print("error: third argument must be a file name")
            usage(sys.argv[0])
            sys.exit(1)

        keyname = sys.argv[2]
        filename = sys.argv[3]
        print("Signing " + filename + " ...")
        prv = pyhsslms.HssLmsPrivateKey(keyname)
        if prv.signFile(filename):
            print("   ... Success. Signature saved in " + filename + ".sig")
        else:
            print("   ... Failed!")

    if sys.argv[1] == 'verify':
        if len(sys.argv) < 3:
            print("error: second argument must be a keyname")
            usage(sys.argv[0])
            sys.exit(1)

        if len(sys.argv) < 4:
            print("error: third argument must be a file name")
            usage(sys.argv[0])
            sys.exit(1)

        keyname = sys.argv[2]
        filename = sys.argv[3]
        pub = pyhsslms.HssLmsPublicKey(keyname)
        if pub.verifyFile(filename):
            print("Signature in " + filename + ".sig is valid.")
        else:
            print("Signature verification failed!")

    if sys.argv[1] == 'showprv':
        if len(sys.argv) < 3:
            print("error: second argument must be a keyname")
            usage(sys.argv[0])
            sys.exit(1)

        keyname = sys.argv[2]
        prv = pyhsslms.HssLmsPrivateKey(keyname)
        print("Private Key: " + keyname + ".prv")
        print(prv.hss_prv.prettyPrint())

    if sys.argv[1] == 'showpub':
        if len(sys.argv) < 3:
            print("error: second argument must be a keyname")
            usage(sys.argv[0])
            sys.exit(1)

        keyname = sys.argv[2]
        pub = pyhsslms.HssLmsPublicKey(keyname)
        print("Public Key: " + keyname + ".pub")
        print(pub.hss_pub.prettyPrint())

    if sys.argv[1] == 'showsig':
        if len(sys.argv) < 3:
            print("error: second argument must be a file name")
            usage(sys.argv[0])
            sys.exit(1)

        filename = sys.argv[2]
        sig = pyhsslms.HssLmsSignature(filename)
        print("Signature: " + filename + ".sig")
        print(sig.hss_sig.prettyPrint())

if __name__ == "__main__":
    main()
